File: ict_zoho_recruit/api/RoleProfile.py
import frappe

def is_empty_html(value):
    return not value or value.strip() in (
        '<div class="ql-editor read-mode"><p><br></p></div>',
        '<div class="ql-editor read-mode"><p></p></div>',
    )

@frappe.whitelist(allow_guest=True)
def get_role_profile_details(role_profile):
    description = None
    custom_requirements = None
    designation = None
    custom_benefits = None
    skills = []

    current_role_profile = role_profile

    while current_role_profile:
        role = frappe.db.get_value(
            "Role Profile",
            current_role_profile,
            [
                "description",
                "custom_requirements",
                "custom_benefits",
                "parent_role_profile",
                "designation",
            ],
            as_dict=True,
        )

        if not role:
            break

        if not description and not is_empty_html(role.description):
            description = role.description

        if not custom_requirements and not is_empty_html(role.custom_requirements):
            custom_requirements = role.custom_requirements

        custom_benefits = custom_benefits or role.custom_benefits
        designation = designation or role.designation

        if not skills:
            skills = frappe.get_all(
                "Designation Skill",
                filters={
                    "parent": current_role_profile,
                    "parenttype": "Role Profile",
                    "parentfield": "skills",
                },
                pluck="skill",
                order_by="idx asc",
            )

        if description and custom_requirements and designation and skills:
            break

        current_role_profile = role.parent_role_profile

    return {
        "designation": designation,
        "description": description ,
        "custom_requirements": custom_requirements,
        "custom_benefits":custom_benefits if  custom_benefits else frappe.db.get_single_value("Zoho Recruit Settings","default_job_post_benefits"),
        "skills": skills,
    }

File: ict_zoho_recruit/public/js/role_profile.js
const set_custom_btns = (frm) => {
    frm.add_custom_button("Create Zoho Job Opening", async () => {
        try {
            const profileDetails = await frappe.xcall(
                "ict_zoho_recruit.api.RoleProfile.get_role_profile_details",
                {
                    role_profile: frm.doc.name
                }
            ) || {};

            const default_benefits = await frappe.db.get_single_value("Zoho Recruit Settings","default_job_post_benefits");
            const target_date_days = await frappe.db.get_single_value("Zoho Recruit Settings","default_target_date_range");

            const target_date = frappe.datetime.add_days(
                frappe.datetime.get_today(),
                target_date_days || 15
            );


            frappe.new_doc("Zoho Job Openings", {
                role_profile: frm.doc.name,
                posting_title: frm.doc.designation || profileDetails.designation || "",
                department_name: frm.doc.custom_department || "",
                number_of_positions: 1,
                date_opened: frappe.datetime.get_today(),
                target_date: target_date,
                job_type: frm.doc.custom_employment_type || "",
                industry: frm.doc.custom_industry_type || "",
                salary: frm.doc.custom_salary || 0,
                work_experience: frm.doc.custom_work_experience || "",
                job_opening_status: "In-progress",
                benefits: frm.doc.custom_benefits || default_benefits || "",
                job_description: frm.doc.description || profileDetails.description || "",
                requirements: frm.doc.custom_requirements || profileDetails.custom_requirements || "",
                title: profileDetails.designation || "",
                job_category: frm.doc.job_category || "",
                posting_title: frm.doc.posting_title || "",
                job_type: "Full-time",
                target_date: target_date,
                skills: (frm.doc.skills || []).map(row => ({
                        skill: row.skill
                    }))
            });

            frappe.show_alert({
                message: __("Zoho Job Opening created. Please review and save."),
                indicator: "green"
            }, 5);

        } catch (error) {
            console.error(
                "Failed to create Zoho Job Opening:",
                error
            );

            frappe.msgprint({
                title: __("Error"),
                indicator: "red",
                message: __(
                    "An unexpected error occurred. Check console for details."
                )
            });
        }
    });
};


frappe.ui.form.on("Role Profile", {
    refresh(frm) {
        set_custom_btns(frm);
    }
});
